built_value_generator/lib/src/serializer_source_class.dart

```diff
@@ -217,26 +217,45 @@ abstract class SerializerSourceClass
   CompilationUnitElement get compilationUnit =>
       element.library.definingCompilationUnit;
 
+  /// Returns the serializer class name for the generated implementation. If the manually
+  /// maintained class is private then we ignore the underscore here, to avoid
+  /// returning a class name starting `_$_`.
+  @memoized
+  String get serializerImplName {
+    final unhidden = name.startsWith('_') ? name.substring(1) : name;
+    return '_\$${unhidden}Serializer';
+  }
+
+  /// Returns the serializer instance name for the generated implementation,
+  /// referenced by the user as a `static serializer` attribute.
+  ///
+  /// If the manually maintained class is private then we ignore the underscore here, to avoid
+  /// returning a class name starting `_$_`.
+  @memoized
+  String get serializerInstanceName {
+    final unhidden = name.startsWith('_') ? name.substring(1) : name;
+    final camelCaseName = _toCamelCase(unhidden);
+    return '_\$${camelCaseName}Serializer';
+  }
+
   Iterable<String> computeErrors() {
     var result = <String>[];
 
     if (!serializerSettings.custom) {
-      final camelCaseName =
-          name.substring(0, 1).toLowerCase() + name.substring(1);
-
       final expectedSerializerDeclaration =
-          'static Serializer<$genericName> get serializer => '
-          '_\$${camelCaseName}Serializer;';
+          'static Serializer<$genericName> get serializer => ${serializerInstanceName};';
       // We used to recommend raw types; recommend the full type now, but still
       // allow raw types.
       final expectedSerializerDeclarationRaw =
-          'static Serializer<$name> get serializer => '
-          '_\$${camelCaseName}Serializer;';
+          'static Serializer<$name> get serializer => ${serializerInstanceName};';
       if (serializerDeclaration != expectedSerializerDeclaration &&
           serializerDeclaration != expectedSerializerDeclarationRaw) {
         result.add('Declare $name.serializer as: '
             '$expectedSerializerDeclaration got $serializerDeclaration');
       }
+      if (name.startsWith('_')) {
+        result.add('Cannot generate serializers for private class ${name}');
+      }
```

Comment:
Looks like you can revert most of the changes to this class, since with this check generation for private classes is no longer needed?

Reply:
I could, but I don't really see the advantage – except maybe clarifying the doc comments. IMO leaving the `_` handling makes it more consistent with the rest of the code base, but with an explicit decision to disallow something, rather than a check to prevent something unimplemented.

We _can_ do this but we _aren't_ vs we _could_ do this but we _haven't_.
Matter of taste though – we can revert the changes if you prefer.

Reply:
I think somewhere in the middle is best :)

Extracting `serializerImplName` and `serializerInstanceName` is an improvement.

But, the `name.startsWith` check on line 256 guarantees that the corresponding checks in those two methods will always return false in any real situation, making that codepath unused and untested. Untested code is better removed if it can be :)

Thanks!

Reply:
@davidmorgan alright, cleaned up the dead code and rebased against master 👍